### PR TITLE
[16.0][MIG] hr_timesheet_time_restriction: Migration to 16.0

### PR DESCRIPTION
**Pull request by tendil**
Standart migration.

### CODE REVIEW

**Review by Bearnard21 — APPROVED** [review]
LGTM

File: hr_timesheet_time_restriction/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2022 Dinar Gabbasov
+# Copyright 2022 Ooops404
+# Copyright 2022 Cetmix
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/__manifest__.py:3]: I have impression that this is OCA module. Should then copyright be as per convention for OCA?

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                days = (
+                    record.project_id.timesheet_restriction_days
+                    or timesheet_restriction_days
+                )
+                if days and today - relativedelta(days=days) > record.date:
+                    raise ValidationError(
+                        _(
+                            "You cannot set a timesheet more than"
+                            " {days} days from current date.".format(
+                                days=record.project_id.timesheet_restriction_days
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py]: Why not use `days` value here? This message can be vague because what we have in lines 46-49.

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                            ),
+                        )
+                    )
+                if (
+                    record.project_id.timesheet_restriction_days
+                    == timesheet_restriction_days
+                    == 0
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py]: Is it more relevant here if `days` equal to 0, not to re-evaluate `record.project_id.timesheet_restriction_days` and `timesheet_restriction_days` again because it is already done in the `days` variable definition?

File: hr_timesheet_time_restriction/tests/test_enabled_hr_timesheet_time_restriction.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                "account_id": self.analytic_account.id,
+                "employee_id": self.employee.id,
+                "name": "Test line",
+            }
+        )
+        self.assertTrue(line, "Timesheet should be created")
+        # check that we can create new timesheet with date before
+        # that current date - 1
+        line = False
+        try:
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/tests/test_enabled_hr_timesheet_time_restriction.py]: This is duplicate code fragment with lines 102-114. Consider pulling it out to setUp if it is not suitable for setUpClass.

File: hr_timesheet_time_restriction/tests/test_disabled_hr_timesheet_time_restriction.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                "project_id": self.project.id,
+                "account_id": self.analytic_account.id,
+                "employee_id": self.employee.id,
+                "name": "Test line",
+            }
+        )
+        self.assertTrue(line, "Timesheet should be created for today")
+        # check that we can create new timesheet with date before
+        # that current date - 1
+        line = self.env["account.analytic.line"].create(
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/tests/test_disabled_hr_timesheet_time_restriction.py]: Duplicate code fragment with lines 128-137, 168-177. Consider pulling it out to setUp or setUpClass depending to business logic.

File: hr_timesheet_time_restriction/models/project_project.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    timesheet_restriction_days = fields.Integer(
+        default=0, help="Not active if equal to 0."
+    )
+
+    use_timesheet_restriction = fields.Boolean(
+        default=lambda self: True
+        if self.env["ir.config_parameter"]
+        .sudo()
+        .get_param("hr_timesheet_time_restriction.use_timesheet_restriction", False)
+        else False
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/project_project.py]: We do not need "if" when the `use_timesheet_restriction` is Boolean by itself.

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            .sudo()
+            .get_param(
+                "hr_timesheet_time_restriction.timesheet_restriction_days",
+                0,
+            )
+        )
+        today = date.today()
+        use_timesheet_restriction = (
+            True
+            if self.env["ir.config_parameter"]
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py]: We do not need "if" when the `use_timesheet_restriction` is Boolean by itself.

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    def _check_project_date(self):
+        """
+        Check that the date can be changed for the project.
+        """
+        if self.user_has_groups(
+            "hr_timesheet_time_restriction.group_timesheet_time_manager"
+        ):
+            return True
+
+        timesheet_restriction_days = int(
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py]: `timesheet_restriction_days` parameter is already integer.

File: hr_timesheet_time_restriction/views/project_project_view.xml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            >
+                <div>
+                    <label
+                        for="use_timesheet_restriction"
+                        string="Use timesheet restriction"
+                        class="oe_inline"
+                    />
+                    <field
+                        name="use_timesheet_restriction"
+                        class="oe_inline oe_input_align"
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/views/project_project_view.xml]: These css classes are not used in v16.

**Review by bnisevic — CHANGES_REQUESTED** [review]

**Review by geomer198 — APPROVED** [review]

File: hr_timesheet_time_restriction/__manifest__.py (hunk 1 above)
Comment (tendil) [reply hr_timesheet_time_restriction/__manifest__.py:3]: The migration documentation says not to change the authors and year in Copyright, so we can remove that if we get a task to remove Copyright. 

https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0

**Review by BhaveshHeliconia — APPROVED** [review]
Functional Tests Ok !! But I think code needs to be improve like mentioned by @bnisevic

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            record (Record): The timesheet record containing the date to be validated.
+            allowed_days (int): The number of days allowed for backdating timesheets. If set
+                to 0, only today's date is permissible.
+
+        Raises:
+            ValidationError: If the record date is older than allowed_days or does not match
+                today's date when allowed_days is set to 0.
+        """
+        today = date.today()
+        delta = today - record.date
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py:102]: Add explicit validation to ensure record.date is always valid before performing operations on it. Additionally, validate allowed_days to ensure it is non-negative.

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        use_global, global_days = self._read_global_restriction()
+
+        for record in self:
+            if not (record.date and record.project_id):
+                continue
+
+            if not (use_global or record.project_id.use_timesheet_restriction):
+                continue
+
+            allowed_days = record.project_id.timesheet_restriction_days or global_days
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py]: global_days could default to 0, is that OK for the functionality?

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            return
+
+        use_global, global_days = self._read_global_restriction()
+
+        for record in self:
+            if not (record.date and record.project_id):
+                continue
+
+            if not (use_global or record.project_id.use_timesheet_restriction):
+                continue
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py]: Why not avoid using continue  and return in the spaghetti manner and write something like:

```
if not self._is_time_manager():
    use_global, global_days = self._read_global_restriction()

    for record in self:
        if record.date and record.project_id and (use_global or record.project_id.use_timesheet_restriction):
            allowed_days = record.project_id.timesheet_restriction_days or global_days
            self._validate_timesheet_date(record, allowed_days)


```

This would be more readable and maintainable.

File: hr_timesheet_time_restriction/tests/test_disabled_hr_timesheet_time_restriction.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        except ValidationError:
+            self.project.timesheet_restriction_days = 0
+        self.assertEqual(self.project.timesheet_restriction_days, -1)
+
+    def test_set_negative_config_restriction_days(self):
+        ConfigSettings = self.env["res.config.settings"].create({})
+        ConfigSettings.timesheet_restriction_days = -1
+        ConfigSettings._onchange_timesheet_restriction_days()
+        ConfigSettings.execute()
+        self.assertEqual(ConfigSettings.timesheet_restriction_days, 0)
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/tests/test_disabled_hr_timesheet_time_restriction.py:133]: Tests do not cover edge cases like:

- Invalid or missing system parameters.
- Scenarios where both global and project-level restrictions are disabled.

File: hr_timesheet_time_restriction/models/account_analytic_line.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+            allowed_days = record.project_id.timesheet_restriction_days or global_days
+
+            self._validate_timesheet_date(record, allowed_days)
+
+    def _is_time_manager(self):
+        """Return True if the current user is in the timesheet time manager group."""
```
Comment (bnisevic) [inline hr_timesheet_time_restriction/models/account_analytic_line.py:45]: It would be useful to log with the logger an unauthorized attempt here.

**Review by bnisevic — CHANGES_REQUESTED** [review]

File: hr_timesheet_time_restriction/models/account_analytic_line.py (hunk 11 above)
Comment (tendil) [reply hr_timesheet_time_restriction/models/account_analytic_line.py]: Yes that's normal.
Nothing was broken during functional testing and the tests pass successfully as well.
The logic of this module has not been changed since version 14

**Review by bnisevic — APPROVED** [review]

**Review by bnisevic — APPROVED** [review]

**Comment by ivs-cetmix** [comment]
@tendil thank you for your contribution! Please squash the administrative commits following the migration Wiki guidelines.
image

**Comment by ivs-cetmix** [comment]
Hey @OCA/project-service-maintainers would appreciate your review and merge!

**Comment by ivs-cetmix** [comment]
Thank you @dreispt !
